thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp

```diff
@@ -0,0 +1,259 @@
+    command_interfaces_config.names.emplace_back(temp_name.str() + "/" + hardware_interface::HW_IF_EFFORT);
+  }
+
+  return command_interfaces_config;
+}
+
+controller_interface::CallbackReturn ThrusterAllocationMatrixController::on_cleanup(
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn ThrusterAllocationMatrixController::on_configure(
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  auto ret = configure_parameters();
+  if (ret != controller_interface::CallbackReturn::SUCCESS) {
+    return ret;
+  }
+
+  // Initialize the reference and state realtime messages
+  const std::shared_ptr<geometry_msgs::msg::Wrench> reference_msg = std::make_shared<geometry_msgs::msg::Wrench>();
+  reference_.writeFromNonRT(reference_msg);
+
+  // Subscribe to the reference topic
+  reference_sub_ = get_node()->create_subscription<geometry_msgs::msg::Wrench>(
+    "~/reference", rclcpp::SystemDefaultsQoS(),
+    [this](const std::shared_ptr<geometry_msgs::msg::Wrench> msg) { reference_callback(msg); });  // NOLINT
+
+  // Setup the controller state publisher
+  controller_state_pub_ =
+    get_node()->create_publisher<control_msgs::msg::MultiDOFStateStamped>("~/status", rclcpp::SystemDefaultsQoS());
+  rt_controller_state_pub_ =
+    std::make_unique<realtime_tools::RealtimePublisher<control_msgs::msg::MultiDOFStateStamped>>(controller_state_pub_);
+
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn ThrusterAllocationMatrixController::on_activate(
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  reference_interfaces_.assign(reference_interfaces_.size(), std::numeric_limits<double>::quiet_NaN());
+
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn ThrusterAllocationMatrixController::on_deactivate(
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::return_type ThrusterAllocationMatrixController::update_and_write_commands(
+  const rclcpp::Time & /*time*/, const rclcpp::Duration & /*period*/)
+{
+  // getting the reference forces and calculating the thrust forces using the tam
+  const std::vector<double> reference_forces_values(reference_interfaces_.begin() + dof_, reference_interfaces_.end());
+  auto reference_forces = create_six_dof_eigen_from_named_vector(dof_names_, six_dof_names_, reference_forces_values);
+
+  Eigen::VectorXd thruster_forces = tam_.completeOrthogonalDecomposition().pseudoInverse() * reference_forces;
+
+  // setting command interfaces with calculated thruster forces
+  for (std::int64_t i = 0; i < num_thrusters_; i++) {
+    command_interfaces_[i].set_value(thruster_forces[i]);
+  }
+
+  // TODO: we still aren't sure if this is the correct type/way of doing this. we may need to come back and do
```

Comment:
Add this back in before you submit the PR

Reply:
Noted.